orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -270,10 +277,12 @@ def forward(self, *inputs, **kwargs):
         new_input_shape = [list(input.size()) for input in inputs if input is not None]
         if self._current_input_shape is None or self._current_input_shape != new_input_shape:
             self._current_input_shape = new_input_shape
-            self._split_training_graph(*inputs, **kwargs)
-        elif self._device_changed:
+            self._build_training_graph()
             self._create_training_session()
-            self._device_changed = False
+        # TODO: disabled for now, since it caused a bug in NVBert fp32 run
+        # elif self._device_changed:
+        #     self._create_training_session()
+        #     self._device_changed = False
```

Comment:
```suggestion
            self._device_changed = False # TODO: disabled for now, since it caused a bug in NVBert fp32 run
        elif self._device_changed:
            self._create_training_session()
            self._device_changed = False
```
That would probably be a cleaner temporary fix

Reply:
I actually prefer to make this explicit. 
I had discussed with Baiju before about this... Supporting changing device requires more careful design. 
For example, if user call .cuda(), .cpu() and then .cuda() in a sequence, without invoking forward(). so there is actually no need to recreate the session, since the start device is the same as end device.